### PR TITLE
[CARBONDATA-3501] Fix update table with varchar column

### PR DESCRIPTION
**Pull request by kevinjmh**
**Problem**
Update on table with varchar column will throw exception:
```
java.lang.Exception: Dataload failed, String length cannot exceed 32000 characters
	at org.apache.carbondata.streaming.parser.FieldConverter$.objectToString(FieldConverter.scala:52)
	at org.apache.carbondata.spark.util.CarbonScalaUtil$.getString(CarbonScalaUtil.scala:70)
	at org.apache.carbondata.spark.rdd.NewRddIterator$$anonfun$next$1.apply$mcVI$sp(NewCarbonDataLoadRDD.scala:357)
	at scala.collection.immutable.Range.foreach$mVc$sp(Range.scala:160)
	at org.apache.carbondata.spark.rdd.NewRddIterator.next(NewCarbonDataLoadRDD.scala:356)
	at org.apache.carbondata.spark.rdd.NewRddIterator.next(NewCarbonDataLoadRDD.scala:333)
	at org.apache.carbondata.processing.loading.steps.InputProcessorStepImpl$InputProcessorIterator.getBatch(InputProcessorStepImpl.java:232)
```

**Analyse**
In the loading part of update operation, it gets the `isVarcharTypeMapping` for each column in the order when table created. And this gives a hint for checking string length. It does not allow to exceeds 32000 char for a column which is not varchar type.

However when changing the plan for updating in `CarbonIUDRule`, it first deletes the old expression and *appends* the new one, which makes the order differ to table created. Such that the string length checking fail.

**Solution**
Keep the order as table created when modify update plan

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/568/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/570/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/574/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/569/

File: integration/spark2/src/main/scala/org/apache/spark/sql/optimizer/CarbonIUDRule.scala [hunk 1]
```diff
@@ -37,21 +39,30 @@ class CarbonIUDRule extends Rule[LogicalPlan] with PredicateHelper {
...
             }
+            // check updated columns exists in table
+            val diff = cols.diff(dest.map(_.name.toLowerCase))
             if (diff.nonEmpty) {
               sys.error(s"Unknown column(s) ${ diff.mkString(",") } in table ${ table.tableName }")
             }
+            // modify plan for updated column *in place*
             isTransformed = true
-            Project(dest.filter(a => !cols.contains(a.name.toLowerCase)) ++ source, child)
+            source.foreach{ col =>
```
Comment (zzcclp) [inline integration/spark2/src/main/scala/org/apache/spark/sql/optimizer/CarbonIUDRule.scala]: add a blank before '{'

**Comment by zzcclp** [comment]
LGTM, test this pr successfully in local.

File: integration/spark2/src/main/scala/org/apache/spark/sql/optimizer/CarbonIUDRule.scala (hunk 1 above)
Comment (kevinjmh) [reply integration/spark2/src/main/scala/org/apache/spark/sql/optimizer/CarbonIUDRule.scala]: fixed

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/585/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/588/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/590/

**Comment by zzcclp** [comment]
@jackylk please help to review, thanks.

**Comment by jackylk** [comment]
LGTM

**Pull request closed without merging**